src/gui/sdlmain.cpp

```diff
@@ -1658,17 +1658,33 @@ void GFX_EndUpdate( const Bit16u *changedLines ) {
 		return;
 	bool actually_updating = sdl.updating;
 	sdl.updating=false;
+	static int64_t last_called = GetTicksUs() - 100000; // ensure first
+	                                                    // frame renders
+	static bool render_current_frame = true;
+	const auto timestamp = GetTicksUs();
+	const auto time_since_last_render = GetTicksDiff(timestamp, last_called);
+	const auto render_skip_threshold = (time_since_last_render / 2);
+	last_called = timestamp;
 	switch (sdl.desktop.type) {
-	case SCREEN_TEXTURE:
+	case SCREEN_TEXTURE: {
 		assert(sdl.texture.input_surface);
-		SDL_UpdateTexture(sdl.texture.texture,
-		                  nullptr, // update entire texture
-		                  sdl.texture.input_surface->pixels,
-		                  sdl.texture.input_surface->pitch);
-		SDL_RenderClear(sdl.renderer);
-		SDL_RenderCopy(sdl.renderer, sdl.texture.texture, NULL, &sdl.clip);
-		SDL_RenderPresent(sdl.renderer);
-		break;
+		if (render_current_frame) {
+			SDL_UpdateTexture(sdl.texture.texture,
+			                  nullptr, // update entire texture
+			                  sdl.texture.input_surface->pixels,
+			                  sdl.texture.input_surface->pitch);
+			SDL_RenderClear(sdl.renderer);
+			SDL_RenderCopy(sdl.renderer, sdl.texture.texture,
+			               nullptr, &sdl.clip);
+			const auto start = GetTicksUs();
+			SDL_RenderPresent(sdl.renderer);
+			const auto elapsed = GetTicksUsSince(start);
+			render_current_frame = elapsed < render_skip_threshold;
+			// LOG_MSG("SDL_RenderPresent took %dus", elapsed);
+		} else {
+			render_current_frame = true;
+		}
```

Comment:
This block of logic is repeated in the three rendering areas - and mixes with values and `static` variables further up. Would be nice to have the overall logic closer together, and not repeated (if possible).

Reply:
I tried wrapping the entire switch statement, but that caused several rendering problems with opengl, and I didn't want to start mangling the code too much.